### ISSUE TITLE
Set rules don't work within show rules except for text

### ISSUE THREAD

**Issue by Scarabytes** [typst/typst#4591]
### Description

I'm new to Typst, so I may have overlooked something. The following report is based on what I'd expect to happen vs what actually happens.

In the "[Making a Template](https://typst.app/docs/tutorial/making-a-template/)" tutorial, an idiom is shown where a #show rule is created for the document that sets up a few #set and #show rules and then forwards the document object that was passed as an argument. 

Using this for custom show rules for text works as expected, turning the text red in this case:
```typst
#let textmod(t) = {
  set text(fill: red)
  t
}

#show text: textmod

#text("hello")
world
```

However, when trying to use the exact same construct to style a table (or several other constructs such as #box or #grid), nothing happens
```typst
#let tablemod(tab) = {
  set table(stroke: none)
  tab
}

#show table: tablemod

#table(
  columns: 2,
  table.header[A][B],
  [a], [b],
)
```

I'm not sure if I'm misunderstanding how it _should_ work, but to me this seems inconsistent and therefore wrong. I'd expect both cases to behave identically.

### Reproduction URL

https://typst.app/project/rkDxp9Uh0GpUipFjvA_sPo

### Operating system

Web app

### Typst version

- [X] I am using the latest version of Typst

**Comment by JamesxX** [comment]
In both cases the `show` part isn't needed - just do a set rule

**Comment by Scarabytes** [comment]
Yes, just using a set rule in the main scope would achieve the result I'm looking for.
However, the ultimate goal is to create a transformer function that does more than just change one default - working on this, I stumbled upon this inconsistency.

Speaking of, I found one more thing where I'm not sure if I'm just misunderstanding the intended behaviour or if it is actually a bug:
Calling the transformer function directly on the table object _does_ remove the stroke. 
```typst
#tablemod(table(
  columns: 2,
  table.header[A][B],
  [a], [b],
  [#sym.alpha], [#sym.beta],
  table.footer[1][2]
))
```
However, I was under the impression that a
```typst
#show table: tablemod
```
has the effect of calling tablemod() on every subsequent occurence of a table, so this and the code in the original message _should_ be equivalent in my understanding. Yet they aren't, so I am confused about whether this is a bug or something I don't understand yet.

**Comment by laurmaedje** [comment]
The thing that differs between the "Making a template" docs and your example is the part before the colon. In your case, there is no need to write `show table:`, just `show:` suffices since that then applies the set rule to the full remaining content in the scope/file, just as if you had written the top-level set rule.

The reason why the `show table` approach doesn't work is because by the time the function runs on `show table`, the table is already fully prepared and cannot be affected by set rules anymore. What _would_ work is `show table: set table(stroke: none)`. In this case, it's useless, but something like `show heading.where(level: 1): set heading(numbering: "I")` can make sense.

The decision that `show x: set x(..)` behaves differently from `show x: it => { set x(..); it }` was a trade-off.[^1] Ideally, both could behave the same. But the problem is that when there are multiple show rules of the latter kind, they need to be evaluated in some order. This means the table goes through various stages of styling and there isn't one canonical representation of it. This has various bad consequences. In contrast, with show-set Typst can statically resolve all rules at once when "preparing" the table. The resulting finished table is canonical and also what you get from `query`.

Note that `show text: ..` isn't useful[^2]: In `show text: set ..` the only sensible right hand-side is `set text(..)` because text contains no other elements, but `show text: set text(..)` doesn't give you anything over `set text(..)`. Meanwhile `show text: it => ..` doesn't really give you anything over `show regex(".."): ..`.

[^1]: Before we [made this change](https://github.com/typst/typst/pull/3311) in 0.11, both behaved exactly the same. That was the source of various bugs. First of all, it was impossible to overwrite show-set rules with subsequent ones. Secondly, show-set rules on the same element behaved very unpredictably, sometimes working and sometimes not, due to various interactions.

[^2]: I have actually not yet come across a _single_ good use case.

**Comment by Scarabytes** [comment]
Thank you so much for the detailled explanation @laurmaedje ! It makes sense that it's much more efficient to pass around full objects rather than a lot of unevaluated context, and I assume the fact that they can't be edited in show rules is part of the core memory model. I want to dive into the Rust code eventually to learn the details about it :) 

It would maybe be worth mentioning the differences of these cases somewhat prominently in the documentation (maybe in a table outlining in which cases this idiom is possible or not) because it _is_ somewhat confusing to be honest. 

But as a learning experience for me: So in general to just set defaults, a set rule is the best bet and if additional logic is needed, either special functions or show rules that reconstruct a new object are the way to go? Or am I overlooking a good third option still?

**Comment by laurmaedje** [comment]
> So in general to just set defaults, a set rule is the best bet and if additional logic is needed that is more complex that where() selectors, either special functions or show rules that reconstruct a new object are the way to go? Or am I overlooking a good third option still?

Yes, set rule or show-set with where selectors is best. A show rule that reconstructs a new objects should only be used as a last resort because it ruins the semantics of the document. When you reconstruct a heading, for instance, it shows up twice in the outline. That's also the reason why we decided to not do anything about show rules applying recursively when doing reconstruction. It is a problematic idiom and we don't want to encourage it. Sometimes there is simply no other option, but it can also indicate that the thing one wants to do is problematic in the first place.

> It would maybe be worth mentioning the differences of these cases somewhat prominently in the documentation (maybe in a table outlining in which cases this idiom is possible or not) because it is somewhat confusing to be honest.

We can repurpose this issue for that.

**Comment by KZNS** [comment]
Besides the difference between `show xxx: set xxx` and `show xxx: it => {}`,the variations between `show xxx: it => {}` and `show: rest => {}` are also something confused me.

According to the output of `repr`, `show xxx: it => {}` creates a `styled` block. It doesn't actually "transform" the elements as the description in the doc.
However, `show: rest => {}` dose transform the elements into a new structure. It only applies on the `sequence` of the rest elements, not on "every" element in the rest, although it looks like `show xxx:` without selector (or with a selector accept all) in the doc.
(Maybe these different behaviors of `show` should be assign to different keywords in the future.)

I think it will be awesome to have a section to describing the core model of Typst, explaining when and how the keywords like `show` will affect it. This would be helpful for the user who want to make some deeper customizations.

**Comment by laurmaedje** [comment]
> Maybe these different behaviors of show should be assign to different keywords in the future.

The bare `show:` actually used to be called `wrap`. But even though the implementation of both is different, conceptually the same thing is happening, hence the unification. Either individual elements or the remaining block is transformed. Just that in the case of bare `show:`, the transformation is eager (since we can apply it to all of the rest directly) whereas for `show xxx:`, it is lazy and only happens as we recurse into the elements in a later compilation stage.

> I think it will be awesome to have a section to describing the core model of Typst, explaining when and how the keywords like show will affect it. This would be helpful for the user who want to make some deeper customizations.

100%
